External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h

```diff
@@ -0,0 +1,507 @@
+                ClientFlags Expected = ClientFlags::FLAG_DISOWNED;
+                if ((*it)->CurrentClientOwnedFlag->compare_exchange_strong(Expected, ClientFlags::FLAG_FREE)) {
+                  // We managed to take away ownership
+                  // Put it back in the regular pool and come back to it
+                  (*it)->CurrentClientOwnedFlag = nullptr;
+                  OwnedBuffers.emplace_back(*it);
+                  it = NonOwnedBuffers.erase(it);
+                  continue;
+                }
+              }
+            }
+
+            ++it;
+          }
+        }
+
+        {
+          // Walk all the allocations and find a buffer that fits
+          for (auto it = OwnedBuffers.begin(); it != BuffersEnd; ++it) {
+            if ((*it)->Size == Size) {
+              BestFit = it;
+              break;
+            }
+
+            if ((*it)->Size > Size) {
+              UnsizedFit = it;
+            }
+          }
+
+          // If we didn't have an exact fit then use an unsized fit
+          if (BestFit == BuffersEnd) {
+            BestFit = UnsizedFit;
+          }
+
+          {
+            std::chrono::time_point<ClockType> LRUTime{};
+            ContainerType::iterator LastUsed = BuffersEnd;
+
+            // Walk all the allocations and find a buffer to erase
+            for (auto it = OwnedBuffers.begin(); it != OwnedBuffers.end(); ++it) {
+              // Ensure that the LRU value is past our duration threshold and isn't the one we are claiming
+              // Also only select a single memory region
+              if (it != BestFit) {
+                auto UsedTime = (*it)->LastUsed.load(std::memory_order_relaxed);
+                if ((Now - UsedTime) >= DURATION &&
+                    UsedTime > LRUTime) {
+                  LastUsed = it;
+                  LRUTime = UsedTime;
+                }
+              }
+            }
+
+            // If we found a buffer then free it
+            if (LastUsed != BuffersEnd) {
+              Free((*LastUsed)->Ptr, (*LastUsed)->Size);
+              delete *LastUsed;
+              OwnedBuffers.erase(LastUsed);
+            }
+          }
+
+          if (BestFit != OwnedBuffers.end()) {
+            MemoryBuffer *Buffer = *BestFit;
+            OwnedBuffers.erase(BestFit);
+            return NonOwnedBuffers.emplace(NonOwnedBuffers.end(), Buffer);
+          }
+        }
+
+        // Need to allocate a new buffer, couldn't fit
+        auto Data = Alloc(Size);
+        return NonOwnedBuffers.emplace(NonOwnedBuffers.end(), new MemoryBuffer{Data, Size, ClockType::now()});
+      }
+
+      void UnclaimBufferImpl(ContainerType::iterator Buffer) {
+        (*Buffer)->CurrentClientOwnedFlag = nullptr;
+        OwnedBuffers.emplace_back(*Buffer);
+        NonOwnedBuffers.erase(Buffer);
+      }
+
+      /**
+       * @brief List of buffers that this pool allocator itself owns
+       */
+      ContainerType OwnedBuffers;
```

Comment:
How about renaming to `UnclaimedBuffers`,  and `ClaimedBuffers` so that `ClaimBuffer`/`ClaimBufferImpl` & `UnclaimBuffer`/`UnclaimBufferImpl` are more intuitive to understand?

`DisownBuffer` and `ReownOrClaimBuffer` don't modify those lists, and from my understanding they track the `claimed`/`unclaimed` statuses, not `owned`/`disowned`

Reply:
Resolved.